main.c

```diff
@@ -1051,6 +1051,12 @@ int __attribute__((used)) main(void) {
     alarm_reset();
     #endif
 
+    #if CIRCUITPY_DISPLAYIO && CIRCUITPY_OS_GETENV && CIRCUITPY_SET_DISPLAY_LIMIT
+    // If number of displays has been overridden in settings.toml allocate memory and
+    // move existing displays
+    malloc_display_memory();
```

Comment:
Since you are doing this really early, can you just have one dynamically allocated array? It'll make all of the other accessing code simpler.

Reply:
That was my original attempt, however the boards that had displays defined as board.DISPLAY didn't like the dynamic memory allocation. I tried multiple pointer configurations to try and get the compile to work but didn't look into modifying how board.DISPLAY was defined.

I can go back and dig into the board.DISPLAY setup and try and get it to work with the dynamically allocated memory structure if you think that makes sense.

Reply:
Ah, right. board.DISPLAY is static.

I wonder if it'd be more fruitful to just try and allocate displays to the CP heap first and then move them over if needed.

Reply:
@tannewt I thought you mentioned somewhere else about replacing `board.DISPLAY` with something else in 10 or 11? Would `board.display()` or similar be helpful?

Reply:
I still don't have a working understanding of the complete CP memory model so I'm not sure how I would do that, is there a mechanism like port_malloc that would allocate heap space?

Reply:
We now have `supervisor.runtime.display` too: https://github.com/adafruit/circuitpython/issues/8675

Right now it is read-only. The longer-term idea was to opt a display into being the console by setting that value instead of doing it implicitly like we do today.

Reply:
Once this PR was set, I was going to take a run at getting the supervisor.runtime.display object to be settable but I have no idea what road blocks I might run into. It looked to me like getting the primary_display variable working was doable and the first step....

Reply:
> I wonder if it'd be more fruitful to just try and allocate displays to the CP heap

m_malloc seems to complain (crash) if the VM is not running. I could allocate the m_malloc memory when the display is allocated in the displayio modules `allocate_display` and `allocate_display_bus`. I think I'd need to add a new structure to keep track of which displays had not yet had memory allocated as well. None of the dynamic displays would survive the VM but I could copy the primary display to the static memory before the VM shut down.

Is that what you had in mind?

Reply:
> None of the dynamic displays would survive the VM but I could copy the primary display to the static memory before the VM shut down.

Yup, that's what I had in mind. I'd only expect it to be called from user code. displayio.DISPLAY and the picodvi init will use the single static allocation that is the display we also use for the terminal.

Reply:
Well the user code displays seem to be working with m_malloc now. I've held off working on moving displays to the static storage as I think that will be related to the primary display features and I thought it would be better to work on that in a follow up PR.

Reply:
Sounds good! Can you have the first display use the single static allocation?

Reply:
You mean if it hasn't been used by picodvi, board.DISPLAY, or anything else yet. Yea, if I do that it will save allocating the dynamic storage for a single user display. I'll give it a go :grin:

Reply:
Actually maybe I'm misunderstanding, I think it already grabs the static storage first.